packages/ERTP/src/paymentLedger.js

```diff
@@ -0,0 +1,267 @@
+  allegedName,
+  brand,
+  assetKind,
+  displayInfo,
+) => {
+  /** @type {WeakStore<Payment, Amount>} */
+  const paymentLedger = makeWeakStore('payment');
+
+  /** @type {(left: Amount, right: Amount) => Amount } */
+  const add = (left, right) => AmountMath.add(left, right, brand);
+  /** @type {(left: Amount, right: Amount) => Amount } */
+  const subtract = (left, right) => AmountMath.subtract(left, right, brand);
+  /** @type {(allegedAmount: Amount) => Amount} */
+  const coerce = allegedAmount => AmountMath.coerce(brand, allegedAmount);
+  /** @type {(left: Amount, right: Amount) => boolean } */
+  const isEqual = (left, right) => AmountMath.isEqual(left, right, brand);
+
+  /** @type {Amount} */
+  const emptyAmount = AmountMath.makeEmpty(brand, assetKind);
+
+  const makePayment = makePaymentMaker(allegedName, brand);
+
+  // Methods like deposit() have an optional second parameter `amount`
+  // which, if present, is supposed to be equal to the balance of the
+  // payment. This helper function does that check.
+  const assertAmountEqual = (paymentBalance, amount) => {
+    if (amount !== undefined) {
+      assert(
+        isEqual(amount, paymentBalance),
+        X`payment balance ${paymentBalance} must equal amount ${amount}`,
+      );
+    }
+  };
+
+  const assertKnownPayment = payment => {
+    assert(paymentLedger.has(payment), X`payment not found for ${allegedName}`);
+  };
+
+  /**
+   * Reallocate assets from the `payments` passed in to new payments
+   * created and returned, with balances from `newPaymentBalances`.
+   * Enforces that total assets are conserved.
+   *
+   * Note that this is not the only operation that reallocates assets.
+   * `purse.deposit` and `purse.withdraw` move assets between a purse and
+   * a payment, and so must also enforce conservation there.
+   *
+   * @param {Payment[]} payments
+   * @param {Amount[]} newPaymentBalances
+   * @returns {Payment[]}
+   */
+  const reallocate = (payments, newPaymentBalances) => {
+    // There may be zero, one, or many payments as input to
+    // reallocate. We want to protect against someone passing in
+    // what appears to be multiple payments that turn out to actually
+    // be the same payment (an aliasing issue). The `combine` method
+    // legitimately needs to take in multiple payments, but we don't
+    // need to pay the costs of protecting against aliasing for the
+    // other uses.
+
+    if (payments.length > 1) {
+      const antiAliasingStore = makeWeakStore('payment');
+      payments.forEach(payment => {
+        if (antiAliasingStore.has(payment)) {
+          throw Error('same payment seen twice');
+        }
+        antiAliasingStore.init(payment, undefined);
+      });
+    }
+
+    const total = payments.map(paymentLedger.get).reduce(add, emptyAmount);
+
+    const newTotal = newPaymentBalances.reduce(add, emptyAmount);
+
+    // Invariant check
+    assert(isEqual(total, newTotal), 'rights were not conserved');
```

Comment:
```suggestion
    assert(isEqual(total, newTotal), X`rights were not conserved: ${total} vs ${newTotal}`);
```
Usually good to be generous with the diagnostic info, given that the extra info is already redacted against prying eyes.

Reply:
Done!